avionics/common/src/roar/buffer.cpp

```diff
@@ -0,0 +1,238 @@
+    : data_(new uint8_t[data_capacity]),
+      subpkts_(new Buffer::DataIt[subpkt_capacity + 1]), pkt_size_(pkt_size),
+      data_cap_(data_capacity), subpkt_cap_(subpkt_capacity + 1),
+      subpkt_begin_(0), subpkt_count_(0), pos_(0) {
+    assert(subpkt_capacity > 0);
+    // can't have more subpackets than bytes
+    assert(data_capacity >= subpkt_capacity);
+    assert(pkt_size > 0);
+
+    subpkts_[0] = Buffer::DataIt{0};
+}
+
+int Buffer::usage() {
+    // Check for zero prevents ambiguity from case where buffer is full
+    if (subpkt_count_ == 0) {
+        return 0;
+    }
+
+    DataIt end_of_last = deref(addIt(subpkt_begin_, subpkt_count_));
+    return subIt(end_of_last, deref(subpkt_begin_));
+}
+
+/**
+ * @brief Returns true if no data is in the buffer.
+ */
+bool Buffer::empty() { return subpkt_count_ == 0; }
+
+/**
+ * @brief Allocates space for a subpacket of specified `size`. Future calls
+ * to write() will be done for the allocated subpacket.
+ *
+ * It is expected that the subpacket gets filled out immediately, before
+ * space for the next subpacket is allocated.
+ *
+ * @param size Number of bytes to allocate
+ * @return Number of subpackets that were dropped to make space
+ */
+int Buffer::allocSubpkt(int size) {
+    // In addition to not making sense, a subpacket with size 0 screws with
+    // computing data usage of the buffer
+    assert(size > 0);
+    assert(size <= pkt_size_);
+
+    // Ensure there's space in the subpacket buffer
+    int starting_subpkt_count = subpkt_count_;
+    // recall: subpkt_cap_ = 1 + [# of subpackets that can be buffered]
+    if (subpkt_count_ + 1 >= subpkt_cap_) {
+        popSubpkt();
+    }
+
+    // Ensure there's space in the data buffer
+    while (usage() + size > data_cap_) {
+        popSubpkt();
+    }
+    int drop_count = starting_subpkt_count - subpkt_count_;
+
+    // Get end of last subpacket already allocated
+    const SubpktIt subpkts_end = addIt(subpkt_begin_, subpkt_count_);
+
+    // previous usage of write() should've brought pos_ to the right place,
+    // since otherwise the subpacket wouldn't be filled
+    assert(pos_ == deref(subpkts_end));
+
+    // but it's safest/more robust to enforce it anyways
+    // Reset data pointer to newest subpacket (for use in write())
+    pos_ = deref(subpkts_end);
+
+    // Mark end of newly allocated subpacket
+    deref(addIt(subpkts_end, 1)) = addIt(pos_, size);
+    ++subpkt_count_;
+
+    return drop_count;
+}
+
+void Buffer::write(uint8_t byte) {
+    deref(pos_) = byte;
+    pos_ = addIt(pos_, 1);
+}
+
+void Buffer::write(uint8_t *dat, int len) {
+    int space_remain = data_cap_ - pos_.index;
+    if (len <= space_remain) {
+        // fits within remaining space
+        std::memcpy(&data_[pos_.index], dat, len);
+    } else {
+        // data wraps to start
+        std::memcpy(&data_[pos_.index], dat, space_remain);
+        std::memcpy(data_.get(), dat + space_remain, len - space_remain);
+    }
```

Comment:
Is this only able to wrap once at a time?

Also does this wrapping result in partially overwritten packets or does addIt take care of that?

Reply:
If it needs to wrap more than once then the subpacket size is larger than the actual buffer size, which violates a key underlying assumption. Although I guess it'd be a good idea to assert that.

Reply:
Yeah at least an assert would be good.